### PR TITLE
[Flow] Add missing export types to source/*

### PR DESCRIPTION
**Pull request by stepankuzmin**
A part of #11426. Add more export types to a `source/*`

### CODE REVIEW

**Review by mourner — CHANGES_REQUESTED** [review]
Needs a lint fix (probably unused `FeatureStates` type in `source_cache.js`).

File: src/source/query_features.js [hunk 1]
```diff
@@ -54,9 +62,18 @@ export function queryRenderedFeatures(sourceCache: SourceCache,
...
         result[layerID].forEach((featureWrapper) => {
-            const feature = featureWrapper.feature;
+            const feature: GeoJSONFeature & {
+                state: FeatureStates;
+                source: string;
+                sourceLayer: string;
+                layer: LayerSpecification;
+            } = (featureWrapper.feature: any);
+
+            // $FlowFixMe prop-missing
```
Comment (stepankuzmin) [inline src/source/query_features.js:72]: `background` and `sky` layers don't have a source, but skipping this step breaks `test-query` suite, ignoring for now

File: src/source/source_cache.js [hunk 2]
```diff
@@ -330,10 +332,10 @@ class SourceCache extends Evented {
      * @private
      */
     _retainLoadedChildren(
-        idealTiles: {[_: any]: OverscaledTileID},
+        idealTiles: {[number | string]: OverscaledTileID},
         zoom: number,
         maxCoveringZoom: number,
-        retain: {[_: any]: OverscaledTileID}
+        retain: {[number | string]: OverscaledTileID}
```
Comment (mourner) [inline src/source/source_cache.js:338]: Nit: let's stick to one type for a key if possible (`tile.key` is `number` so I think we can do away with `number`).

**Review by mourner — CHANGES_REQUESTED** [review]

File: src/source/source_cache.js (hunk 2 above)
Comment (stepankuzmin) [reply src/source/source_cache.js:338]: Flow can't infer a number type as a key and there are many iterations over `this._tiles` https://flow.org/try/#0PTAEAEDMBsHsHcBQiDGsB2BnALqbBLaAU0wC5QBvAbXQFcBbAIyICcBdcgQ3QE8BfUAF5KfZJFgtQACjRZcufOjyESASkqJQW0LJzLiQ-SSrY2AbkSigA

```js
const tiles: {[number]: any} = {}

for (const t in tiles) {
    const tile = tiles[t];
}
```

```
const tile = tiles[t];
                   ^ string [1] is incompatible with number [2]. [incompatible-type]
References:
5: for (const t in tiles) {
              ^ [1]
3: const tiles: {[number]: any} = {}
                  ^ [2]
```

**Review by mourner — APPROVED** [review]
🎉

**Pull request merged**